### ISSUE TITLE
Performance issues using C standard

### ISSUE THREAD

**Issue by RisingLiberty** [lefticus/cpp_weekly#201]
**C++ Weekly**

**Topics**

The C standard is widely used and is also used in C++.
However, some of its code, which should be trivial and fast, is actually slow.

example: https://quick-bench.com/q/ZE1DFYD0jP7DN6t9MtOfpzaGH8Y
a simple "toupper" implementation can be 8x faster on gcc to 22x faster on clang than using the C standard version.

it'd be nice to discuss this in an episode and possibly inspire deeper investigation into which functions are slow and why.
I believe they're slow because the compiler can't inline the function calls.

**10 min**

**Comment by christosg88** [comment]
Check the example at cppreference.com [https://en.cppreference.com/w/cpp/string/byte/toupper](toupper).

`std::toupper` takes into account the current locale as well, which your `to_upper` function doesn't.

**Comment by RisingLiberty** [comment]
apologies, I though tstd::toupper didn't check the C locale,
I do have another example though, here's a testcase for std::div that's 12x slower than if you'd implement it yourself.

it's less drastic as the previous example, but there's still a performance difference where a custom implementation is 1.3 times faster.
https://quick-bench.com/q/EVx72v42DMw-wq__71nMv0FmrWc

**Comment by christosg88** [comment]
I looked at the implementation of `std::div` in gcc's standard library, and the `std::div_t` struct returned, and they are exactly like yours.

The only difference I could spot is that the compiler is able to inline your function, but it doesn't inline `std::div`.

I don't know the why, so I'll upvote you.
I'd also like Jason to explain the why.

**Comment by reedhedges** [comment]
Yes I think you're right about inlining, constexpr evaluation and optimization in general.   See https://quick-bench.com/q/sjj2moXIWRyZ3Iw-V80g_978I4U where I tried to prevent those specific optimizations (so it's closer to how the functions are in the C library).     

Same with div(). -- https://quick-bench.com/q/9rzsn8kkXJHpv4HMbE9m7AOsOg4 (note had to change it to GCC for [[gnu::noinline]]).

There are also other hidden performance issues in the C library, especially for strings and global state like locale.    But I imagine something like the math functions are almost as fast as they could be?     What about things like like filesystem API? pthreads? sockets?  So lots to think about.
